### PR TITLE
Fix mono_arch_setup_async_callback and add mono_arch_setup_resume_sighandler_ctx on PPC

### PR DESCRIPTION
**Pull request by NattyNarwhal**
mono_arch_setup_async_callback didn't work on systems that use
function pointers. For consistency with other backends with
exception support, also add mono_arch_setup_resume_sighandler_ctx,
which we can use to support mono_arch_setup_async_callback, like
those other backends.

Also enable async callbacks on AIX, then enable
*resume_sighandler_ctx on PPC Linux (untested) and AIX. With this,
I can now ^C in the REPL with AIX and have it work; whereas before
it bombed out, with or without async callbacks on/off.

### CODE REVIEW

**Comment by lewurm** [comment]
approve

**Comment by lewurm** [comment]
build

**Comment by lewurm** [comment]
@monojenkins build Linux PPC64 little endian

**Comment by lewurm** [comment]
I think this breaks the debugger on PPC: https://github.com/mono/mono/blob/a19fad65f3f62372170eeabcc23553f6b73b81e1/mono/mini/debugger-agent.c#L5163

**Comment by lewurm** [comment]
you can run the debugger tests with:
```sh
$ make -C mcs/class/Mono.Debugger.Soft run-test TEST_HARNESS_VERBOSE=1
```
I tried your PR on ppc64le/linux and it doesn't introduce additional failures _there_, but ppc64le doesn't use function descriptors. Could you give it a try?

There are two tests hard crashing (you will need to disable them, see patch in gist) and seven "regular" failures: https://gist.github.com/lewurm/d6ec9ced5b4c90fdec84c53aa35d5cc8

**Comment by lewurm** [comment]
on linux I get the following warnings:
```
  CC       libmini_la-exceptions-ppc.lo
exceptions-ppc.c: In function ?mono_arch_setup_async_callback?:
exceptions-ppc.c:816:23: warning: assignment makes integer from pointer without a cast [-Wint-conversion]
  *(unsigned long *)sp = MONO_CONTEXT_GET_SP(ctx);
                       ^
exceptions-ppc.c:818:45: warning: passing argument 2 of ?mono_arch_setup_resume_sighandler_ctx? makes pointer from integer without a cast [-Wint-conversion]
  mono_arch_setup_resume_sighandler_ctx(ctx, (unsigned long) async_cb);
                                             ^
In file included from exceptions-ppc.c:31:0:
mini.h:2185:10: note: expected ?gpointer {aka void *}? but argument is of type ?long unsigned int?
 void     mono_arch_setup_resume_sighandler_ctx  (MonoContext *ctx, gpointer func);
          ^
```

File: mono/mini/exceptions-ppc.c [hunk 1]
```diff
@@ -808,9 +808,24 @@ void
 mono_arch_setup_async_callback (MonoContext *ctx, void (*async_cb)(void *fun), gpointer user_data)
 {
 	uintptr_t sp = (uintptr_t) MONO_CONTEXT_GET_SP(ctx);
+	/*
+	 * TODO: set user_data somewhere; I assume first arg greg? s390x uses
+	 * reg[2] and amd64 RDI
+	 */
```
Comment (lewurm) [inline mono/mini/exceptions-ppc.c]:
```c
ctx->regs [PPC_FIRST_ARG_REG] = user_data
```
should do it?

**Comment by NattyNarwhal** [comment]
Sorry, was horribly busy today.

I had tried with and patch the patch, and some light tweaks to the patch, but the debugger/runtime seem to crash on the same test regardless. (I haven't tested modifying the test suite to nop those tests out though.) It doesn't seem like we'd be failing on tests any worse; especially considering I don't think it would have done anything different in the end anyways. (Maybe that function descriptor twiddler in `debugger-agent` can be removed if resume context is aware anyways.)

Good point on the register setting, I'll try to add that.

**Comment by NattyNarwhal** [comment]
```
Stacktrace:


=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================


Tests run: 89, Passed: 80, Errors: 2, Failures: 7, Inconclusive: 0
  Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
Elapsed time: 00:06:27.0360000

Errors and Failures:

1) Bug59649 (MonoTests.DebuggerTests.Bug59649)
     Expected string length 4 but was 8. Strings differ at index 0.
  Expected: "Call"
  But was:  "Bug59649"
  -----------^

  at MonoTests.DebuggerTests.assert_location (Mono.Debugger.Soft.Event e, System.String method) [0x0000e] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:537
  at MonoTests.DebuggerTests.Bug59649 () [0x00036] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:4438
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


2) Exceptions (MonoTests.DebuggerTests.Exceptions)
     Expected string length 22 but was 9. Strings differ at index 0.
  Expected: "NullReferenceException"
  But was:  "Exception"
  -----------^

  at MonoTests.DebuggerTests.Exceptions () [0x00255] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:3277
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


3) LineNumbers (MonoTests.DebuggerTests.LineNumbers)
     Expected: 1832
  But was:  1831

  at MonoTests.DebuggerTests.LineNumbers () [0x0009c] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:2413
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


4) Locals (MonoTests.DebuggerTests.Locals)
     Expected: 0
  But was:  22

  at MonoTests.DebuggerTests.AssertValue (System.Object expected, System.Object val) [0x00080] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:330
  at MonoTests.DebuggerTests.Locals () [0x003c3] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:2251
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


5) SetIP (MonoTests.DebuggerTests.SetIP)
   Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
TearDown : Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x0011d] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1566
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1573
  at Mono.Debugger.Soft.Connection.VM_Resume () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1670
  at Mono.Debugger.Soft.VirtualMachine.Resume () [0x00006] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs:125
  at MonoTests.DebuggerTests.run_until (System.String name) [0x00034] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:138
  at MonoTests.DebuggerTests.SetIP () [0x00093] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:4207
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
--TearDown
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x0011d] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1566
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1573
  at Mono.Debugger.Soft.Connection.VM_Resume () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1670
  at Mono.Debugger.Soft.VirtualMachine.Resume () [0x00006] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs:125
  at MonoTests.DebuggerTests.TearDown () [0x00022] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:348
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


6) SetIPSingleStep (MonoTests.DebuggerTests.SetIPSingleStep)
     Expected: 3
  But was:  2

  at MonoTests.DebuggerTests.AssertValue (System.Object expected, System.Object val) [0x00080] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:330
  at MonoTests.DebuggerTests.SetIPSingleStep () [0x000d3] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:4243
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


7) SingleStepping (MonoTests.DebuggerTests.SingleStepping)
     Expected string length 15 but was 3. Strings differ at index 1.
  Expected: "single_stepping"
  But was:  "ss2"
  ------------^

  at MonoTests.DebuggerTests.assert_location (Mono.Debugger.Soft.Event e, System.String method) [0x0000e] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:537
  at MonoTests.DebuggerTests.SingleStepping () [0x00055] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:643
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


8) String_GetValue (MonoTests.DebuggerTests.String_GetValue)
     String lengths are both 2. Strings differ at index 1.
  Expected: "\0A"
  But was:  "\0䄀"
  -------------^

  at MonoTests.DebuggerTests.String_GetValue () [0x00022] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:3883
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


9) VTypes (MonoTests.DebuggerTests.VTypes)
   Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
TearDown : Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x00117] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1565
  at Mono.Debugger.Soft.Connection.StackFrame_GetThis (System.Int64 thread_id, System.Int64 id) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:2394
  at Mono.Debugger.Soft.StackFrame.GetThis () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/StackFrame.cs:178
  at MonoTests.DebuggerTests.VTypes () [0x004c9] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:1985
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
--TearDown
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x0011d] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1566
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1573
  at Mono.Debugger.Soft.Connection.VM_Resume () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1670
  at Mono.Debugger.Soft.VirtualMachine.Resume () [0x00006] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs:125
  at MonoTests.DebuggerTests.TearDown () [0x00022] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:348
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


Results saved as /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/TestResult-net_4_x.xml.
```

**Comment by NattyNarwhal** [comment]
Looks to be the same for the PR:

```
Stacktrace:


=================================================================
Got a SIGSEGV while executing native code. This usually indicates
a fatal error in the mono runtime or one of the native libraries
used by your application.
=================================================================


Tests run: 89, Passed: 80, Errors: 2, Failures: 7, Inconclusive: 0
  Not run: 0, Invalid: 0, Ignored: 0, Skipped: 0
Elapsed time: 00:06:27.0890000

Errors and Failures:

1) Bug59649 (MonoTests.DebuggerTests.Bug59649)
     Expected string length 4 but was 8. Strings differ at index 0.
  Expected: "Call"
  But was:  "Bug59649"
  -----------^

  at MonoTests.DebuggerTests.assert_location (Mono.Debugger.Soft.Event e, System.String method) [0x0000e] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:537
  at MonoTests.DebuggerTests.Bug59649 () [0x00036] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:4438
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


2) Exceptions (MonoTests.DebuggerTests.Exceptions)
     Expected string length 22 but was 9. Strings differ at index 0.
  Expected: "NullReferenceException"
  But was:  "Exception"
  -----------^

  at MonoTests.DebuggerTests.Exceptions () [0x00255] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:3277
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


3) LineNumbers (MonoTests.DebuggerTests.LineNumbers)
     Expected: 1832
  But was:  1831

  at MonoTests.DebuggerTests.LineNumbers () [0x0009c] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:2413
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


4) Locals (MonoTests.DebuggerTests.Locals)
     Expected: 0
  But was:  22

  at MonoTests.DebuggerTests.AssertValue (System.Object expected, System.Object val) [0x00080] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:330
  at MonoTests.DebuggerTests.Locals () [0x003c3] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:2251
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


5) SetIP (MonoTests.DebuggerTests.SetIP)
   Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
TearDown : Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x0011d] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1566
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1573
  at Mono.Debugger.Soft.Connection.VM_Resume () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1670
  at Mono.Debugger.Soft.VirtualMachine.Resume () [0x00006] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs:125
  at MonoTests.DebuggerTests.run_until (System.String name) [0x00034] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:138
  at MonoTests.DebuggerTests.SetIP () [0x00093] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:4207
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
--TearDown
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x0011d] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1566
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1573
  at Mono.Debugger.Soft.Connection.VM_Resume () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1670
  at Mono.Debugger.Soft.VirtualMachine.Resume () [0x00006] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs:125
  at MonoTests.DebuggerTests.TearDown () [0x00022] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:348
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


6) SetIPSingleStep (MonoTests.DebuggerTests.SetIPSingleStep)
     Expected: 3
  But was:  2

  at MonoTests.DebuggerTests.AssertValue (System.Object expected, System.Object val) [0x00080] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:330
  at MonoTests.DebuggerTests.SetIPSingleStep () [0x000d3] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:4243
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


7) SingleStepping (MonoTests.DebuggerTests.SingleStepping)
     Expected string length 15 but was 3. Strings differ at index 1.
  Expected: "single_stepping"
  But was:  "ss2"
  ------------^

  at MonoTests.DebuggerTests.assert_location (Mono.Debugger.Soft.Event e, System.String method) [0x0000e] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:537
  at MonoTests.DebuggerTests.SingleStepping () [0x00055] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:643
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


8) String_GetValue (MonoTests.DebuggerTests.String_GetValue)
     String lengths are both 2. Strings differ at index 1.
  Expected: "\0A"
  But was:  "\0䄀"
  -------------^

  at MonoTests.DebuggerTests.String_GetValue () [0x00022] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:3883
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


9) VTypes (MonoTests.DebuggerTests.VTypes)
   Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
TearDown : Mono.Debugger.Soft.VMDisconnectedException : Exception of type 'Mono.Debugger.Soft.VMDisconnectedException' was thrown.
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x00117] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1565
  at Mono.Debugger.Soft.Connection.StackFrame_GetThis (System.Int64 thread_id, System.Int64 id) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:2394
  at Mono.Debugger.Soft.StackFrame.GetThis () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/StackFrame.cs:178
  at MonoTests.DebuggerTests.VTypes () [0x004c9] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:1985
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305
--TearDown
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command, Mono.Debugger.Soft.Connection+PacketWriter packet) [0x0011d] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1566
  at Mono.Debugger.Soft.Connection.SendReceive (Mono.Debugger.Soft.Connection+CommandSet command_set, System.Int32 command) [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1573
  at Mono.Debugger.Soft.Connection.VM_Resume () [0x00000] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/Connection.cs:1670
  at Mono.Debugger.Soft.VirtualMachine.Resume () [0x00006] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/VirtualMachine.cs:125
  at MonoTests.DebuggerTests.TearDown () [0x00022] in /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/Test/dtest.cs:348
  at (wrapper managed-to-native) System.Reflection.MonoMethod.InternalInvoke(System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00038] in /opt/monodev/mono/mcs/class/corlib/System.Reflection/MonoMethod.cs:305


Results saved as /opt/monodev/mono/mcs/class/Mono.Debugger.Soft/TestResult-net_4_x.xml.
gmake[1]: *** [../../build/tests.make:208: run-test-lib] Error 1
gmake[1]: Leaving directory '/opt/monodev/mono/mcs/class/Mono.Debugger.Soft'
gmake: *** [../../build/rules.make:208: do-run-test] Error 1
gmake: Leaving directory '/opt/monodev/mono/mcs/class/Mono.Debugger.Soft'
```

I'll try adding a patch or two on top.

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -5167,7 +5167,11 @@ resume_from_signal_handler (void *sigctx, void *func)
 #endif
 	mono_monoctx_to_sigctx (&ctx, sigctx);
 
-#ifdef PPC_USES_FUNCTION_DESCRIPTOR
+#if defined(PPC_USES_FUNCTION_DESCRIPTOR) && !defined(MONO_ARCH_HAVE_SETUP_RESUME_FROM_SIGNAL_HANDLER_CTX)
```
Comment (lewurm) [inline mono/mini/debugger-agent.c]: is this ever true?

I think that can be removed, including `mono_ppc_set_func_into_sigctx`

Comment (NattyNarwhal) [reply mono/mini/debugger-agent.c]: I assume it's a theoretical case.... but you're right I guess; anyone porting Mono to a new OS on PPC can just make sure it runs fine with `MONO_ARCH_HAVE_SETUP_RESUME_FROM_SIGNAL_HANDLER_CTX)`.

Still unsure about the Mac OS case, but I don't think Mac OS uses function descriptors in any case.

Comment (lewurm) [reply mono/mini/debugger-agent.c]: I think it's safe to define `MONO_ARCH_HAVE_SETUP_RESUME_FROM_SIGNAL_HANDLER_CTX` for mac OS too.

**Comment by lewurm** [comment]
`mono_ppc_set_func_into_sigctx` is unused now, please remove it from `mono/mini/mini-ppc.h` and `mono/mini/exceptions-ppc.c` 🙂

**Comment by NattyNarwhal** [comment]
Already taken care of, just building before I push :)

**Comment by NattyNarwhal** [comment]
Done. Only case I'm slightly concerned about now is the PPC Mac OS one, and I'm pretty sure the #ifdef labyrinth in place with this patch should effectively handle it.

**Comment by lewurm** [comment]
one more request: let's define `MONO_ARCH_HAVE_SETUP_RESUME_FROM_SIGNAL_HANDLER_CTX` for every platform. I don't see any reason why it wouldn't work on Apple.

**Comment by NattyNarwhal** [comment]
`MONO_ARCH_HAVE_SIGCTX_TO_MONOCTX` too?

**Comment by lewurm** [comment]
yeah, `MONO_ARCH_HAVE_SIGCTX_TO_MONOCTX` would be fine for every platform too. It's not related to your PR really, so it's up to you 😉

**Comment by NattyNarwhal** [comment]
Might want to merge before more conflicts come in in :)

**Pull request merged**